### PR TITLE
Remmediate [PIH-01M] Inexistent Support of Manual Seller Configuration

### LINKED ISSUE #728: [PIH-01M] Inexistent Support of Manual Seller Configuration
## PIH-01M: Inexistent Support of Manual Seller Configuration

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Language Specific](/appendix/finding-types#language-specific) |  | [ProtocolInitializationHandlerFacet.sol:L172-L178](https://github.com/bosonprotocol/boson-protocol-contracts/blob/02a4d2ff04832a4607674ba37708d4f52cdac4fa/contracts/protocol/facets/ProtocolInitializationHandlerFacet.sol#L172-L178) |

### Description:

The [`ProtocolInitializationHandlerFacet::initV2_3_0`](https://github.com/bosonprotocol/boson-protocol-contracts/blob/02a4d2ff04832a4607674ba37708d4f52cdac4fa/contracts/protocol/facets/ProtocolInitializationHandlerFacet.sol#L153-L182) function will assign the `sellerCreator` values of multiple seller IDs to ensure that existing deployed seller IDs remain compatible with the updated system, however, this type of upgrade mechanism can only process a finite number of seller IDs and can become inexecutable in the future.

### Impact:

The current v2.3.0 upgrade initialization methodology may become unsustainable as the protocol grows due to the inherent block gas limits that are present on all blockchains.

### Example:

```solidity
for (uint256 i = 0; i < sellerIds.length; i++) {
    (bool exists, , ) = fetchSeller(sellerIds[i]);
    require(exists, NO_SUCH_SELLER);
    require(sellerCreators[i] != address(0), INVALID_ADDRESS);

    lookups.sellerCreator[sellerIds[i]] = sellerCreators[i];
}
```

### Recommendation:

We advise the code to instead allow `sellerCreator` values to be filled in either via administrative action in batches, or by the owner of a particular `BosonVoucher` instance, either of which we consider an adequate remediation to this exhibit.

### PR DESCRIPTION
**Pull request by zajck**
Closes #728 

Based on #746 which should be reviewed and merged first.

Instead of backfilling the data at the initialization time, seller salts are now generated 
- when a seller is created for new sellers
- dynamically the first time the new collection is created for existing sellers.
   - Existing sellers can also manually set it by calling `updateSellerSalt`

Due to the changes in #746 it will be possible that the same wallet creates multiple seller ids. It's also possible that already now there exist seller accounts with the same creator address. To avoid collisions, seller salt is now derived from the admin address and the user-supplied salt.
Since this is a user-defined value, we cannot just pick it on the sellers' behalf.

Two view methods were added to check if the chosen salt will result in successful clone creation:
- `isSellerSaltAvailable` can be used before calling `createSeller`, `createNewCollection` (for sellers from before v2.3.0 which might not have seller salt initialized yet) or `updateSellerSalt`
- `calculateCollectionAddress` can be used to get both the expected collection address and its availability.

### CODE REVIEW

File: contracts/interfaces/handlers/IBosonAccountHandler.sol [hunk 1]
```diff
@@ -321,6 +323,20 @@ interface IBosonAccountHandler is IBosonAccountEvents {
         BosonTypes.VoucherInitValues calldata _voucherInitValues
     ) external;
 
+    /**
+     * @notice Updates a salt.
+     * Use this if the admin address is updated and there exist a possibility that old admin will try to create the vouchers
```
Comment (albertfolch-redeemeum) [inline contracts/interfaces/handlers/IBosonAccountHandler.sol:328]:
```suggestion
     * Use this if the admin address is updated and there exists a possibility that the old admin will try to create the vouchers
```

File: contracts/protocol/facets/SellerHandlerFacet.sol [hunk 2]
```diff
@@ -380,6 +394,35 @@ contract SellerHandlerFacet is SellerBase {
         emit CollectionCreated(sellerId, collectionIndex, voucherCloneAddress, _externalId, assistant);
     }
 
+    /**
+     * @notice Updates a salt.
+     * Use this if the admin address is updated and there exist a possibility that old admin will try to create the vouchers
```
Comment (albertfolch-redeemeum) [inline contracts/protocol/facets/SellerHandlerFacet.sol:399]:
```suggestion
     * Use this if the admin address is updated and there exists a possibility that the old admin will try to create the vouchers
```

File: contracts/protocol/facets/SellerHandlerFacet.sol [hunk 3]
```diff
@@ -366,11 +366,22 @@ contract SellerHandlerFacet is SellerBase {
         Collection[] storage sellersAdditionalCollections = lookups.additionalCollections[sellerId];
         uint256 collectionIndex = sellersAdditionalCollections.length + 1; // 0 is reserved for the original collection
 
+        bytes32 sellerSalt = lookups.sellerSalt[sellerId];
+
+        // Accounts created before v2.3.0 can be missing sellerSalt, so it's created here
+        if (sellerSalt == 0) {
+            (, Seller storage seller, ) = fetchSeller(sellerId);
+            sellerSalt = keccak256(abi.encodePacked(seller.admin, _voucherInitValues.collectionSalt));
```
[the remark this answers is not in the document]
Comment (zajck) [reply contracts/protocol/facets/SellerHandlerFacet.sol]: Resolved.
Similarly, I had to change `updateSellerSalt`, too.

**Review by albertfolch-redeemeum — DISMISSED** [review]

File: contracts/protocol/facets/SellerHandlerFacet.sol [hunk 4]
```diff
@@ -491,6 +506,64 @@ contract SellerHandlerFacet is SellerBase {
...
+            if (admin == address(0)) {
+                admin = IERC721(lookups.authTokenContracts[authToken.tokenType]).ownerOf(authToken.tokenId);
+            }
+            sellerSalt = keccak256(abi.encodePacked(admin, _collectionSalt));
+        }
+
+        // Calculate collection address
+        bytes32 collectionSalt = keccak256(abi.encodePacked(sellerSalt, _collectionSalt));
+        bytes32 bytecodeHash = keccak256(
+            abi.encodePacked(
```
Comment (levalleux-ludo) [inline contracts/protocol/facets/SellerHandlerFacet.sol:554]: @zajck I'd like you to explain this

Comment (levalleux-ludo) [reply contracts/protocol/facets/SellerHandlerFacet.sol:554]: did I find the key here: https://blog.openzeppelin.com/deep-dive-into-the-minimal-proxy-contract ?

Comment (zajck) [reply contracts/protocol/facets/SellerHandlerFacet.sol:554]: This is directly connected to how all voucher clones are created. Lines 554-558 here represent the minimal clone bytecode that is deployed for each collection, equivalent to how we construct the bytecode in `cloneBosonVoucher`:

https://github.com/bosonprotocol/boson-protocol-contracts/blob/e6d14e4f4dfe8089f4276389a7725023672116a8/contracts/protocol/bases/SellerBase.sol#L177-L188

Then it just needs to be hashed in order to compute create2 address as per the definition.

**Review by levalleux-ludo — APPROVED** [review]

File: contracts/protocol/facets/SellerHandlerFacet.sol [hunk 5]
```diff
@@ -491,6 +506,64 @@ contract SellerHandlerFacet is SellerBase {
         return (pl.cloneAddress[_sellerId], pl.additionalCollections[_sellerId]);
     }
 
+    /**
+     * @notice Returns the availability of salt for a seller.
+     *
+     * @param _adminAddres - the admin address to check
+     * @param _salt - the salt to check (corresponds to `collectionSalt` when `createSeler` or `createNewCollection` is called or `newSalt` when `updateSellerSalt` is called)
```
Comment (albertfolch-redeemeum) [inline contracts/protocol/facets/SellerHandlerFacet.sol:513]:
```suggestion
     * @param _salt - the salt to check (corresponds to `collectionSalt` when `createSeller` or `createNewCollection` is called or `newSalt` when `updateSellerSalt` is called)
```

File: contracts/protocol/facets/SellerHandlerFacet.sol [hunk 6]
```diff
@@ -491,6 +506,64 @@ contract SellerHandlerFacet is SellerBase {
...
+     * @return isAvailable - salt can be used
+     */
+    function isSellerSaltAvailable(address _adminAddres, bytes32 _salt) external view returns (bool isAvailable) {
+        bytes32 sellerSalt = keccak256(abi.encodePacked(_adminAddres, _salt));
+        return !protocolLookups().isUsedSellerSalt[sellerSalt];
+    }
+
+    /**
+     * @notice Calculates the expected collection address and tells if it's still avaialble.
```
Comment (albertfolch-redeemeum) [inline contracts/protocol/facets/SellerHandlerFacet.sol:522]:
```suggestion
     * @notice Calculates the expected collection address and tells if it's still available.
```

File: contracts/protocol/facets/SellerHandlerFacet.sol [hunk 7]
```diff
@@ -532,4 +605,22 @@ contract SellerHandlerFacet is SellerBase {
         (exists, seller, authToken) = fetchSeller(_sellerId);
         seller.clerk = address(0);
     }
+
+    /**
+     * @notice Performs a validation that the message sender is the either the admin address or owner of auth token
```
Comment (albertfolch-redeemeum) [inline contracts/protocol/facets/SellerHandlerFacet.sol:610]:
```suggestion
     * @notice Performs a validation that the message sender is either the admin address or the owner of auth token
```

**Review by albertfolch-redeemeum — APPROVED** [review]
LGTM

**Review by levalleux-ludo — APPROVED** [review]

**Review by albertfolch-redeemeum — APPROVED** [review]

**Pull request merged**